### DISCUSSION TITLE
Query MAM by stanza ID

### DISCUSSION THREAD

**Discussion by Musacca** [processone/ejabberd#3994] [Questions] [answered]
I have a question related to MAM query. So Whenever I send a message, I receive back only a stanza id, however, I cannot query MAM based on stanza id. I can query MAM for a specific user with either message afterUid/beforeUid(which is different from stanza id), or Date. I can get message Uid for received messages but I can't get it for outgoing messages. 

Is there any way to query MAM by stanza ID? If not then, how can I get a message Uid for outgoing message?

**Comment by prefiks** [comment]
You can use it is query rsm:
```
<iq type='set' id='q29302'>
  <query xmlns='urn:xmpp:mam:2'>
    <set xmlns='http://jabber.org/protocol/rsm'>
      <after>{VALUE FROM stanza-id}</after>
    </set>
  </query>
</iq>
```
This will return messages directly after id you provided.

**Reply by Musacca** [reply]
Thanks for your reply.
However, it does not work as expected. It returns all the messages

This is how I send amessage with Smack(Android)
```
                <message to='ab@test.com' id='383941AA-2496-43F3-A395-9C90DBA038FB' type='normal'>
                              <body>
                               Hello
                              </body>
                              <stanza-id xmlns='urn:xmpp:sid:0' id='QNS7-HZX4-ZF1E-S' by='c@test.com'/>
                              <origin-id xmlns='urn:xmpp:sid:0' id='AHUC-FZ2B-P3JK-P'/>          
             </message>
 ```
As you can see, I set the stanza-id element, however, on the receiver side, I get stanza-id different(which is generated by ejabberd server). With the stanza-id I received on the receiver side, I can query MAM and it works perfectly. However, on the sender side, it is not working as expected. 

Is it possible to enable the ejabberd server to take my stanza-id instead of generating a new one on the server side?

**Reply by weiss** [reply, marked as answer]
> Is it possible to enable the ejabberd server to take my stanza-id instead of generating a new one on the server side?

That's currently not supported by the protocol. One issue would be that the server cannot blindly trust client-generated IDs to be unique (how to handle the case where multiple clients use the same IDs, or a single client reuses IDs, and so on).

A pragmatic solution is to just query the history <after/> the last known stanza ID, and then de-duplicate known messages (maybe simply based on your client-generated message ID).

**Reply by Musacca** [reply]
That is how I have implemented it right now. I would like to prevent getting duplicated messages. As I see there is not such an option for that.
